ruby/ql/lib/codeql/ruby/ast/Method.qll

```diff
@@ -47,45 +64,40 @@ class MethodBase extends Callable, BodyStmt, Scope, TMethodBase {
 private class MethodModifier extends MethodCall {
   /** Gets the name of the method that this call applies to. */
   Expr getMethodArgument() { result = this.getArgument(0) }
-
-  /** Holds if this call modifies a method with name `name` in namespace `n`. */
-  pragma[noinline]
-  predicate modifiesMethod(Namespace n, string name) {
-    this = n.getAStmt() and
-    [
-      this.getMethodArgument().getConstantValue().getStringlikeValue(),
-      this.getMethodArgument().(MethodBase).getName()
-    ] = name
-  }
 }
 
-/** A call to `private` or `private_class_method`. */
-private class Private extends MethodModifier {
+/** A method call that sets the visibility of other methods. */
+private class VisibilityModifier extends MethodModifier {
   private Namespace namespace;
   private int position;
 
-  Private() { this.getMethodName() = "private" and namespace.getStmt(position) = this }
+  VisibilityModifier() {
+    this.getMethodName() =
+      [
+        "public", "private", "protected", "public_class_method", "private_class_method",
+        "protected_class_method"
+      ] and
+    namespace.getStmt(position) = this
+  }
 
-  override predicate modifiesMethod(Namespace n, string name) {
-    n = namespace and
-    (
-      // def foo
-      // ...
-      // private :foo
-      super.modifiesMethod(n, name)
-      or
-      // private
-      // ...
-      // def foo
-      not exists(this.getMethodArgument()) and
-      exists(MethodBase m, int i | n.getStmt(i) = m and m.getName() = name and i > position)
-    )
+  /**
+   * Holds if this modifier changes the "ambient" visibility - i.e. the default
+   * visibility of any subsequent method definitions.
+   */
+  predicate modifiesAmbientVisibility() {
+    this.getMethodName() = ["public", "private", "protected"] and
+    this.getNumberOfArguments() = 0
   }
-}
 
-/** A call to `private_class_method`. */
-private class PrivateClassMethod extends MethodModifier {
-  PrivateClassMethod() { this.getMethodName() = "private_class_method" }
+  string getVisibility() {
```

Comment:
Could add a `private newtype Visibility = Public() or Private() or Protected()` and use that instead of `string`.

Reply:
Should be OK if we make `getVisibilityModifier` private as well (`VisibilityModifier` already is).